text/204-golang-bindings.md

```diff
@@ -1027,6 +1030,187 @@ func (c *cluster) HasEc2Capacity() (result bool) {
+```
+
+In order to represent this in Go, we use pointers for all types.
+
+```go
+type OptionlPropertyProps struct {
+  OptionalString *string
+}
+
+type myClassStruct struct {
+}
+
+type MyClass Interface {
+  OptionalString() *string
+}
+
+func NewMyClass(props *OptionalPropertyProps) MyClass {
+  // magic jsii client implementation
+  return &myClassStruct{}
+}
+
+func MyClass_SomeStatic(arg *string) *string {
+  // magic jsii client implementation
+  return arg
+}
+
+// OptionalString is nil
+myStruct := OptionalPropertyProps{}
+myClass := NewMyClass(&myStruct)
+fmt.Println(myClass.OptionalString()) // =>nil
+```
+
+All method arguments, return types, behavioral interface fields, and struct fields need to be pointer types. In cases where a type is a Go interface, ie a jsii
+generated class or behavioral interface, these are already pointer types, as interfaces dictate method signatures for pointer receivers.
+
+```go
+type MyJsiiInterface interface {
+  Property() *string
+}
+
+type MyJsiiStruct struct {
+  IProperty MyJsiiInterface
+}
+```
+
+### Downsides
+
+This is annoying for users when passing simple values to structs or class methods. This can be mitigated by providing the user with convenience functions to
+allow them to continue to define these values inline.
+
+```go
+// You have to allocate a variable here in order to address it via `&`
+myOptionalString := "string"
+myStruct := OptionalPropertyProps{OptionalString: &myOptionalString}
+
+// Now using a helper function with signature string -> *string
+myStruct := OptionalPropertyProps{OptionalString: jsii.String("string")}
+
+type SomeStructSlice {
+  MySlice: *[]*string
+}
+// Complex types can be addressed without allocation as they are already pointers
+myStruct := SomeStructSlice{MySlice: &[]*string{jsii.String("string")}}
+```
+
+Additionally, users will have to always check returned values for nil pointers before dereferencing them.
+
+```go
+myStruct := OptionalPropertyProps{}
+myClass := NewMyClass(&myStruct)
+// I want to access the string value
+fmt.Println(*myClass.OptionalString()) // =>Panics!
+
+// Good and safe
+if myClass.OptionalString() != nil {
+  fmt.Println(*myClass.OptionalString()) // =>Panics!
+}
+```
+
+Essentially every type that the user passes to or receives from the JSII runtime will have to be a pointer type. We can't only make
+optional values pointers because then when a required value became optional in TS, this causes a type change in the generated go code which requires a code
+change. This means many non-breaking changes to the TS source would require major version bumps to the generated go modules.
+
+With these downsides, it still seems like this is the most idiomatic route to take. Go users are familiar with using pointer types in this way from other
+popular modules. Additionally, it is the only strategy we can identify that allows us to maintain JSII's versioning compatibility strategy.
+
+### Explored Alternatives:
```

Comment:
Is there any well-known established Go module that handles null primitives? Maybe leaning on something like that makes sense and then we don't need explicit pointers everywhere, no?

I found this one which looks pretty solid: https://github.com/guregu/null

It's ergonomics are not that bad.

Reply:
Synced offline with the team about this and agreed that the ergonomics look very similarly with wrapper types, via a library or our own, as they do with pointers. Sticking with pointers as the recommendation as this seems to be the precedent in other libraries (gh api and aws-sdk notably) and is more what users expect. Leaving the option open to re-evaluate this in the future based on user feedback. Especially should be re-visited once generics are available and we can provide generalized wrapper types.